Use tags in conduct info

markusjura opened the pull request:
Using tags in the `conduct info` and `conduct info my-bundle` commands. Falling back to the compatibility version if `/bundles` does not include a `tags` key (prior to ConductR 2.1.0) or the tags are empty.

markusjura committed:
Use tags in conduct info

Using tags in the `conduct info` and `conduct info my-bundle` commands. Falling back to the compatibility version if `/bundles` does not include a `tags` key (prior to ConductR 2.1.0) or the tags are empty.

markusjura commented:
## Manual tests

### conduct info - ConductR 2.1.0 - Bundles with tags

```
$ conduct info
Licensed To: cc64df31-ec6b-4e08-bb6b-3216721a56b@lightbend
Max ConductR agents: 3
ConductR Version(s): 0.1.0, 2.1.*
Grants: akka-sbr, cinnamon, conductr

ID               NAME                    TAG  #REP  #STR  #RUN  ROLES
bdfa43d-e5f3504  conductr-haproxy  first-tag     1     0     1  haproxy
3349b6b          eslite            first-tag     1     0     1  elasticsearch
```

### conduct info - ConductR 2.1.0 - no tags in bundles

```
$ conduct info
Licensed To: cc64df31-ec6b-4e08-bb6b-3216721a56b@lightbend
Max ConductR agents: 3
ConductR Version(s): 0.1.0, 2.1.*
Grants: akka-sbr, cinnamon, conductr

ID               NAME              TAG  #REP  #STR  #RUN  ROLES
bdfa43d-e5f3504  conductr-haproxy         1     0     1  haproxy
3349b6b          eslite                   1     0     1  elasticsearch
```

### conduct info - ConductR 2.0.5 - Fallback to compatibility version

```
$ conduct info
ID               NAME              VER  #REP  #STR  #RUN  ROLES
bdfa43d-e5f3504  conductr-haproxy   v2     1     0     1  haproxy
3349b6b          eslite             v1     1     0     1  elasticsearch
```

### conduct info eslite - with tags

```
$ conduct info eslite
BUNDLE ATTRIBUTES
-----------------
Bundle Id              3349b6b
Bundle Name            eslite
Compatibility Version  1
System                 eslite
System Version         1
Tags                   1.0.0, latest
Nr of CPUs             0.1
Memory                 402653184
Disk Space             200000000
Roles                  elasticsearch
Bundle Digest          3349b6bfbc09e8d88123fee5b33b2717e07fb2a91de719c99e0403c9e4b28ea5
Error                  No

BUNDLE SCALE
------------
Nr of Reschedules  0
Scale              1

BUNDLE INSTALLATIONS
--------------------
Host    192.168.10.1
Bundle  /Users/mj/.conductr/images/tmp/conductr/192.168.10.1/bundles/3349b6bfbc09e8d88123fee5b33b2717e07fb2a91de719c99e0403c9e4b28ea5.zip

BUNDLE EXECUTIONS
-----------------
ENDPOINT     HOST          STARTED  BIND_PORT  HOST_PORT
akka-remote  192.168.10.1      Yes      10822      10822
es           192.168.10.1      Yes      10007      10007

SERVICE NAMES
-------------
SERVICE NAME    STATUS
elastic-search  Running
```

### Load latest bundle and bundle configuration from Bintray

```
$ conduct load cassandra cassandra-prod
Retrieving bundle..
Loading bundle from cache typesafe/bundle/cassandra
Bintray credentials loaded from /Users/mj/.lightbend/commercial.credentials
Retrieving file:///Users/mj/workspace/conductr-cli/cassandra
Resolving bundle typesafe/bundle/cassandra
Bintray credentials loaded from /Users/mj/.lightbend/commercial.credentials
Retrieving https://dl.bintray.com/typesafe/bundle/typesafe/cassandra/v3-1acac1d0ce36c7c8824632c658561f85d83a1f7338eb89f5d79dc0baa9118b64/cassandra-v3-1acac1d0ce36c7c8824632c658561f85d83a1f7338eb89f5d79dc0baa9118b64.zip
[#################################################] 100%
Retrieving configuration..
Loading bundle configuration from cache typesafe/bundle-configuration/cassandra-prod
Bintray credentials loaded from /Users/mj/.lightbend/commercial.credentials
Retrieving file:///Users/mj/workspace/conductr-cli/cassandra-prod
Resolving bundle configuration typesafe/bundle-configuration/cassandra-prod
Bintray credentials loaded from /Users/mj/.lightbend/commercial.credentials
Retrieving https://dl.bintray.com/typesafe/bundle-configuration/typesafe/cassandra-prod/v3-3b6fc49c77381436637806f2b866da503b154f58ee77f2b9d16b215cbb72e1ca/cassandra-prod-v3-3b6fc49c77381436637806f2b866da503b154f58ee77f2b9d16b215cbb72e1ca.zip
[#################################################] 100%
Loading bundle to ConductR..
[#################################################] 100%
Bundle 1acac1d0ce36c7c8824632c658561f85-3b6fc49c77381436637806f2b866da50 is installed
Bundle loaded.
Start bundle with:        conduct run 1acac1d-3b6fc49
Unload bundle with:       conduct unload 1acac1d-3b6fc49
Print ConductR info with: conduct info
Print bundle info with:   conduct info 1acac1d-3b6fc49
```

### Loading specific tag from Bintray

```
$ conduct load cassandra:v3
Retrieving bundle..
Loading bundle from cache typesafe/bundle/cassandra:v3
Bintray credentials loaded from /Users/mj/.lightbend/commercial.credentials
Retrieving cassandra:v3
Resolving bundle typesafe/bundle/cassandra:v3
Bintray credentials loaded from /Users/mj/.lightbend/commercial.credentials
Retrieving https://dl.bintray.com/typesafe/bundle/typesafe/cassandra/v3-1acac1d0ce36c7c8824632c658561f85d83a1f7338eb89f5d79dc0baa9118b64/cassandra-v3-1acac1d0ce36c7c8824632c658561f85d83a1f7338eb89f5d79dc0baa9118b64.zip
[#################################################] 100%
Loading bundle to ConductR..
[#################################################] 100%
Bundle 1acac1d0ce36c7c8824632c658561f85 is installed
Bundle loaded.
Start bundle with: conduct run 1acac1d
Unload bundle with: conduct unload 1acac1d
Print ConductR info with: conduct info
```

### Loading current ConductR version from Bintray

```
$ sandbox run 2.0.5 --no-default-features
|------------------------------------------------|
| Starting ConductR                              |
|------------------------------------------------|
Bintray credentials loaded from /Users/mj/.lightbend/commercial.credentials
Retrieving https://dl.bintray.com/lightbend/commercial-releases/conductr-2.0.5-Mac_OS_X-x86_64.tgz
[#################################################] 100%
Bintray credentials loaded from /Users/mj/.lightbend/commercial.credentials
Retrieving https://dl.bintray.com/lightbend/commercial-releases/conductr-agent-2.0.5-Mac_OS_X-x86_64.tgz
[#################################################] 100%
Extracting ConductR core to /Users/mj/.conductr/images/core
Extracting ConductR agent to /Users/mj/.conductr/images/agent
Starting ConductR core instance on 192.168.10.1..
Waiting for ConductR to start..
Starting ConductR agent instance on 192.168.10.1..
|------------------------------------------------|
| Summary                                        |
|------------------------------------------------|
|- - - - - - - - - - - - - - - - - - - - - - - - |
| ConductR                                       |
|- - - - - - - - - - - - - - - - - - - - - - - - |
ConductR has been started:
  core instance on 192.168.10.1
  agent instance on 192.168.10.1
ConductR service locator has been started on:
  192.168.10.1:9008
|- - - - - - - - - - - - - - - - - - - - - - - - |
| Proxy                                          |
|- - - - - - - - - - - - - - - - - - - - - - - - |
HAProxy has not been started
To enable proxying ensure Docker is running and restart the sandbox
|- - - - - - - - - - - - - - - - - - - - - - - - |
| Bundles                                        |
|- - - - - - - - - - - - - - - - - - - - - - - - |
Check latest bundle status with:
  conduct info
Current bundle status:
ID  NAME  TAG  #REP  #STR  #RUN  ROLES
```

markusjura commented:
@huntc This PR does not include the change of `bundle_shorthand.parse_bundle`  because I was running out of time. Also, note that the `bintray_resolver` uses `bundle_shorthand.parse_bundle` and then does some dedicated `compatibility_version` logic: https://github.com/typesafehub/conductr-cli/blob/master/conductr_cli/resolvers/bintray_resolver.py#L195-L268

longshorej commented on conductr_cli/conduct_info_list.py:
```diff
@@ -38,7 +38,7 @@ def display_bundles_default(args, is_license_success, conductr_license, bundles)
         'id': 'ID',
         'name': 'NAME',
-        'compatibility_version': 'VER',
+        'tag': 'TAG',
```
I think it would be more accurate if we show `VER` as the header if `tags` is missing for every bundle in the HTTP response -- in that case, we are in fact showing `compatabilityVersion` so it would be nice if the headers reflect that. wdyt?

huntc replied:
Not a bad idea. If tags is missing for one bundle it'll be missing for all. We could decide between the existing behaviour and tags based on this, and not have the fallback logic as a consequence.

longshorej reviewed (approved):
Looks good. Just had one comment on header name.

huntc reviewed (changes requested):
I'd like to discuss Jason's suggestion and also get the compat version/ shorthand expr done.

markusjura committed:
Use TAG or VER depending if tags key exist

If all HTTP bundle responses contain a `tags` key` then we will display the column name in the `conduct info` command is `TAG` and we will display the tags (could be non existing). Otherwise, we display the compatibility version and name the column `VER`.

markusjura commented:
Incorporated @longshorej review feedback. Now the column is named either `TAG` or `VER` depending if the bundles `tags` key of the HTTP bundle endpoint exists or not.

markusjura committed:
Parsing tag instead of compatibility version

Changes the bundle parsing logic of the `bundle_shorthand` module. When splitting the `tag-digest` string to individual attributes we now tolerate that the tag substring can contain hyphens.

huntc commented:
@markusjura The tests are failing - suggest that you run tests locally first.

huntc reviewed (approved):
One failing test, otherwise LGTM. Please be sure to update our documentation: https://github.com/typesafehub/conductr-doc/blob/master/src/main/play-doc/operation/DeployingBundlesOps.md

huntc commented on conductr_cli/bundle_shorthand.py:
```diff
@@ -46,8 +46,8 @@ def split_package_to_parts(package):
         if '-' in version:
-            compatiblity_version, digest = version.split('-')
-            return package_name, compatiblity_version, digest
+            tag, digest = version.rsplit('-', 1)
```
Nice
